server.py

```diff
@@ -15,9 +15,9 @@
 Question = namedtuple('Question', 'output statement')
 Submission = namedtuple('Submission', 'question time result output')
 
-
 for i in os.listdir(question_dir):
-    with open(os.path.join(question_dir, i, 'output.txt'), 'r') as fl:
+    # read the correct output as bytes object
+    with open(os.path.join(question_dir, i, 'output.txt'), 'rb') as fl:
```

Comment:
Haha! Good catch!

Any reason you picked this vs decoding the user's uploaded data? As I understand it, right now you read the expected file in binary and then when the user's upload comes up, that is also in binary so both can be correctly compared.

The other way would have been to read the expected outputs normally and then when the user uploads something, decode it on the fly.

Any reason for choosing one over the other?

Reply:
Yes! There is a compelling reason to check file equivalency using this approach. It is a generalized method that will work on all sorts of binary files (text, image, video, etc). 

Moreover, if the output.txt contains characters that are unrecognizable by the codec used for decoding a bytes object (`utf-8`, `charmap`, etc), then it will throw `UnicodeDecodeError`. To avoid all this fuss it's better to compare two files in their binary form.

Performance is another reason. It takes time to encode and decode objects, especially when their size is large.

Obviously, there are much better/efficient ways to compare two files for checking result, but this project is in early phase so I don't want to complicate things unless it's required.

Reply:
Just now I checked your GitHub profile and found a similar project named OpenJudge. In that project, you're comparing file content after decoding them.
```Python 
result = result and stdout.decode().strip() == self.out.strip()
```
Any specific reason for using this approach?

Reply:
umm, okay.